vm/src/stdlib/socket.rs

```diff
@@ -691,32 +691,51 @@ fn socket_getnameinfo(
     })
 }
 
-fn get_addr(vm: &VirtualMachine, addr: impl ToSocketAddrs) -> PyResult<socket2::SockAddr> {
-    match addr.to_socket_addrs() {
-        Ok(mut sock_addrs) => {
-            if let Some(mut addr) = sock_addrs.next() {
-                if option_env!("RUSTPYTHON_NO_IPV6").is_some() {
-                    while addr.ip() == IpAddr::V6(Ipv6Addr::LOCALHOST) {
-                        if let Some(other) = sock_addrs.next() {
-                            addr = other
-                        } else {
-                            break;
-                        }
-                    }
-                }
-
-                Ok(addr.into())
-            } else {
-                let error_type = vm.class("_socket", "gaierror");
-                Err(vm.new_exception_msg(
-                    error_type,
-                    "nodename nor servname provided, or not known".to_owned(),
-                ))
-            }
+enum IpAddrFmt {
+    Ipv4(),
+    Ipv6(),
+    Any(),
+}
+
+impl<T> From<T> for IpAddrFmt
+where
+    T: PartialEq<i32> + Sized + std::fmt::Debug,
+{
+    fn from(i: T) -> Self {
+        if i == Domain::ipv4().into() {
+            Self::Ipv4()
+        } else if i == Domain::ipv6().into() {
+            Self::Ipv6()
+        } else {
+            error!("Unknown IP family/domain: {:?}", i);
+            Self::Any()
         }
+    }
+}
+
+fn get_addr<F>(vm: &VirtualMachine, addr: impl ToSocketAddrs, fmt: F) -> PyResult<socket2::SockAddr>
+where
+    F: Into<IpAddrFmt>,
```

Comment:
Maybe the "fmt" argument could just be something like `family: Option<i32>`? I think that's more direct (since this is only used in internal code), and it avoids needing `IpAddrFmt`